Expose StressLog via CDAC and port StressLogAnalyzer to managed code

jkoritzinsky opened the pull request:
# Current state in main

StressLog is an in-memory logging apparatus in CoreCLR and NativeAOT. This log is an in-memory circular log that stores messages as "offsets to format string in memory and array of arguments as `size_t` values".

This format has been generally stable but has evolved slowly over time. Unlike most types in CoreCLR, the StressLog was never DACized. Furthermore, it was never abstracted on the SOS-DAC boundary. As a result, SOS has intimate knowledge of the memory layout and format of the StressLog structures. The StressLog types are one of the reasons the SOSBreakingChangeVersion exists (and are responsible for 2 of the 4 version increases done over the years).

SOS supports the stresslog types by having copies of the stresslog header in its codebase and implementing some of the methods only declared when `STRESS_LOG_READONLY` is defined. The `!dumplog` and `!histinit` and other GC history commands utilize the stress log.

Additionally, dotnet/runtime has a tool called StressLogAnalyzer that supports reading memory-mapped stress logs (which don't require dumps). This code-base has a modified copy of the stress-log dumping/reading code from dotnet/diagnostics (including some dead code and some shims to make things compile) to support dumping and analyzing GC behavior based on stress logs.

# After this PR

With this PR, a StressLog contract is defined in the cDAC. This contract provides 3 versions to enable SOS to use cDAC for all stress log dumping, even for older target runtimes. These versions differ by message format and the lookup process for the format string from the provided offset. Each of these messages corresponds to a different SOS Breaking Change Version. For versions that are supported by memory-mapped stress logs, each version corresponds to a version of the memory-mapped stress log. As there are few versions and they only differ by pointer-size, the contract descriptors for these cases can be hard-coded in SOS if a descriptor is not available.

CoreCLR (and NativeAOT) will implement version 2 of the contract, which is the newest version in this PR.

The StressLog contract exposes the information needed to dump the header of the stress log in SOS, get the logs for all threads, and get all messages from a given thread's log.

The contract does not include an implementation of formatting a stress log message into a string, but it does include a conceptual description of how to do so.

StressLogAnalyzer has been re-written into C# and uses the cDAC reader to read the StressLog. The following behavior differences are noted:

- Specifying `-tid` without a thread while also specifying a thread will not change thread ids to hex. When filtering threads, specify `--hex` to get hex thread ids.
- Specifying `-f` without a format string while also specifying a format string will not output format strings. When specifying format strings, pass `-pf` to print format strings as well. 

# Follow-ups

This PR does not present a mechanism to push a managed cDAC package to the internal feed for SOS to consume. This would be necessary to move SOS to use cDAC to implement StressLog dumping in managed code, or any of the other commands currently implemented in the native portion of SOS.

Once a package exists that SOS can consume, the code to read the StressLog in C++ can be removed from CoreCLR as well. I don't want to remove it before we can remove the copy in SOS.

After this PR is merged, we can update the memory-mapped stresslog format to store a pointer to the contract descriptor in the runtime image. That will ensure that any future changes to the stresslog format won't require versioning the memory-mapped stresslog format version, only updating the contract.

jkoritzinsky committed:
Start a StressLog contract for CDAC

Start on the data parsing and fill out the contract more

Implement stress message payload parsing and thread stress log iteration

Implement everything except for printing the string output

Implement output string formatting for stress log messages

Move all StressMsg logic into the contract and out of the data type.

Split the stress log contract into different versions and refactor out the StressLogHeader to be entirely a StressLogAnalyzer concern (not used in the cDAC implementation at all)

Expose the module table directly to avoid going through the StressLog object to read the module table (which isn't present in the memory-mapped log)

Give a nicer error message for an easy-to-fall-in special case in the data descriptor scraping.

Fix a few datadescriptor problems

jkoritzinsky committed:
Start on StressLogAnalyzer port.

Move stress message formatting out of the contract.

jkoritzinsky committed:
Load in and read the thread stress logs and the stress messages.

jkoritzinsky committed:
Add support for Ix, I64 specifiers

jkoritzinsky committed:
Refactor into src/test directories and add tests for basic format strings

jkoritzinsky committed:
Add more tests and extend the signed integer test case to test a ton more cases (including the alignment scenarios).

jkoritzinsky committed:
Add special pointer formatting hook and a default implementation for StressLogAnalyzer

jkoritzinsky committed:
Write the command line and start work on porting over logic (first one is format string matching).

jkoritzinsky committed:
Add GC heap map memory and default string handling

jkoritzinsky committed:
Implement GC heap map association

jkoritzinsky committed:
Fill in more processing and start on sketching out the output handling.

jkoritzinsky committed:
Port over remaining features

jkoritzinsky committed:
Remove native StressLogAnalyzer

jkoritzinsky committed:
Get 'dump all messages' working correctly

jkoritzinsky committed:
Fix format offset calculation and update contract md to include references to the used data descriptors.

jkoritzinsky committed:
Merge branch 'main' of github.com:dotnet/runtime into cdac-stresslog-analyzer

jkoritzinsky committed:
Add a conceptual description of how to format a StressLog message

jkoritzinsky committed:
Implement escape sequence interpretation

jkoritzinsky committed:
Redefine maxMsgSize to make it C++11 constexpr-compliant even with our debug return logging (make it a const instead of a function)

lambdageek commented on src/coreclr/tools/cdac-build-tool/ObjectFileScraper.cs:
```diff
@@ -387,6 +387,10 @@ private Content ReadContent(ScraperState state, HeaderDirectory header)
         {
+            if (endMagic.All(b => b == 0))
+            {
+                throw new InvalidOperationException("expected endMagic, got all zeros. Did you add something to the data descriptor that can't be initialized at compile time?");
```
What was it?

I wonder if we can do something like this:

```cxx
template <typename T>
T force_constexpr (T arg) constexpr
{
    return arg;
}
```

and then wrap every argument in all the `#define`s in `datadescriptor.cpp`  with `force_constexpr`:

```cxx
        /* .GlobalLiteralValues = */ {
#define CDAC_GLOBAL(name,tyname,value) { /*.Name = */ GET_GLOBAL_NAME(name), /* .TypeName = */ GET_GLOBALTYPE_NAME(name), /* .Value = */ force_constexpr (value) },
#include "datadescriptor.h"
        },

jkoritzinsky replied:
The case that I hit was the `StressLog::maxMsgSize()` function not being constexpr. I ended up moving it to a constexpr field eventually to get the required behavior.

lambdageek replied:
C++20 will eventually give us `consteval` so we can tighten up the datadescriptor.cpp macros.  https://godbolt.org/z/8PG57rrzs

there are some awful hacks to make C++17 templates that would work but i'm not sure it's worth it

jkoritzinsky replied:
We'd still need to update from C++11 first, which is already a struggle.

jkoritzinsky committed:
Merge branch 'main' of github.com:dotnet/runtime into cdac-stresslog-analyzer

jkoritzinsky committed:
Add "is pointer in stresslog" contract function for SOS's GCHandleLeaks functionality

jkoritzinsky committed:
Add some additional members to the contract to allow implementing all SOS StressLog introspection

jkoritzinsky committed:
Wire up ISOSDacInterface::GetStressLogAddress while I'm here

jkoritzinsky committed:
Merge branch 'main' of github.com:dotnet/runtime into cdac-stresslog-analyzer

jkoritzinsky committed:
Merge branch 'main' into cdac-stresslog-analyzer

jkoritzinsky committed:
Merge branch 'main' into cdac-stresslog-analyzer

jkoritzinsky committed:
Update the StressLog contract to better describe and cross-link the custom format specifiers

jkoritzinsky committed:
Start adjusting output to match StressLogAnalyzer

jkoritzinsky committed:
Fix discovering the first message of a new chunk

Maoni0 commented:
please make sure that our dprintf stresslog scenario works, ie, turning on the stress log and analyzing it with the stresslog analyzer. if you have questions, @cshung can (continue to) help.

jkoritzinsky committed:
Fix remaining output nits that would cause the most confusion.

jkoritzinsky committed:
Lots of perf improvements and some fixes

jkoritzinsky committed:
Rewrite GC time index searching for better perf

jkoritzinsky committed:
Make GC thread filtering more aggressive and fix non-GC thread filtering.

Add support for %f because GC uses it.

jkoritzinsky commented:
@Maoni0 I've validated locally (and pushed the necessary fixes to do so) that all of the various options that were supported on the old stress log analyzer (dprintf level filtering, facility ignore, value searching, gc time searching, time searching, etc) are all functioning the same as they were with the old stress log.

I've also added a `--single-run` flag that exits after outputting instead of prompting for new input (which I used in debugging and perf testing).

Maoni0 commented:
nice, thank you @jkoritzinsky! @cshung, could you please also do some validation on your end?

jkoritzinsky committed:
Hook into clr.tools and clr.toolstests subsets

jkoritzinsky committed:
Remove uncessary cast

jkoritzinsky committed:
Correctly implement the quit/retry commands

jkoritzinsky committed:
Merge branch 'main' of github.com:dotnet/runtime into cdac-stresslog-analyzer

jkoritzinsky committed:
Use string helpers from Target

jkoritzinsky committed:
Remove dead code now that StressLogAnalyzer is in C#

jkoritzinsky commented:
Can I get a review pass on this?

@cshung have you validated all of the scenarios you want here before merging?

cshung commented:
> Can I get a review pass on this?
> 
> @cshung have you validated all of the scenarios you want here before merging?

I haven't. I plan to work on it today after doing a code review.
But you don't need to block on me on the initial merge, if I have found anything we can always fix them later.

cshung commented:
@jkoritzinsky WIth the latest commit, it looks like the code no longer compiles if `#define TRACE_GC` is uncommented in `gc.cpp`?

Looks like something is wrong with respect to handling `float`.

jkoritzinsky commented:
> @jkoritzinsky WIth the latest commit, it looks like the code no longer compiles if `#define TRACE_GC` is uncommented in `gc.h`?
> 
> Looks like something is wrong with respect to handling `float`.

That failure is unrelated to this PR (it exists in main). It's due to the fact that `float` and `double` were being handled incorrectly and float can't actually be distinguished from `double` (which was being handled incorrectly as well) in a printf format string.

It looks like a number of failures have crept in under `TRACE_GC`. I'll put out another PR to try to fix them (and maybe add a CI leg to validate the build in the future).

jkoritzinsky commented:
@Maoni0 @cshung I've opened https://github.com/dotnet/runtime/pull/108089 for the fixes for the TRACE_GC build as they're in main and not due to this PR.

jkoritzinsky committed:
Merge remote-tracking branch 'dotnet/main' into cdac-stresslog-analyzer

jkoritzinsky committed:
Update StressLogAnalyzer (not the tests) to the new refactoring.

jkoritzinsky committed:
Fix tets and use Moq to mock the target instead of constructing a fake memory space

jkoritzinsky committed:
Merge branch 'main' of github.com:dotnet/runtime into cdac-stresslog-analyzer

jkoritzinsky committed:
Merge branch 'main' into cdac-stresslog-analyzer

jkoritzinsky committed:
Add support for padded strings as they're apparently used by the GC.

jkoritzinsky commented on docs/design/datacontracts/StressLog.md:
````diff
@@ -0,0 +1,474 @@
+bool IsPointerInStressLog(StressLogData stressLog, TargetPointer pointer);
+```
+
+## Versions 0 to 2
````
I wanted to try to match the StressLog contract versions to the versions of the memory mapped stresslog format (so v1 matches to contract version 1, v2 matches to contract v2). I'm okay renumbering though if you prefer.

jkoritzinsky committed:
Merge branch 'main' of https://github.com/dotnet/runtime into cdac-stresslog-analyzer

jkoritzinsky commented on src/native/managed/cdacreader/Microsoft.Diagnostics.DataContractReader.Contracts/Contracts/StressLog.cs:
```diff
@@ -0,0 +1,387 @@
+        }
+    }
+
+    public bool IsPointerInStressLog(StressLogData stressLog, TargetPointer pointer)
```
This is used by the `!GCHandleLeaks` command in SOS to determine that an appearance of a particular handle isn't a leak, it's just recorded in the stress log.

elinor-fung replied:
Ah, missed that it is in the contract doc. Add it to `IStressLog`?

jkoritzinsky replied:
Will do!

jkoritzinsky committed:
PR feedback and allow a module table to be present or absent in v1/v2 (and remove v0) to support NativeAOT with no changes to the contract

jkoritzinsky committed:
Fix typo

jkoritzinsky commented on src/native/managed/cdacreader/Microsoft.Diagnostics.DataContractReader.Contracts/Contracts/StressLog.cs:
```diff
@@ -0,0 +1,387 @@
+        uint pointerSize = (uint)target.PointerSize;
+
+        ulong payload1 = target.Read<ulong>(msg.Header);
+        ulong payload2 = target.Read<ulong>((ulong)msg.Header + 8);
```
I've included the definitions of the structs here. I think that's the most helpful way to do this as each constant can be directly mapped to the field it's reading.

jkoritzinsky committed:
Update IStressLog.cs

jkoritzinsky committed:
Fix body and accessibility

jkoritzinsky committed:
Make the message filter internal

jkoritzinsky committed:
Merge branch 'main' of https://github.com/dotnet/runtime into cdac-stresslog-analyzer

jkoritzinsky committed:
Move the formatter for stress messages into the Contracts assembly as the same logic will be used from any StressLog reader.

The implementation of the formatter will not be part of the contract itself, but the description of the format string's structure is part of the contract (this is already the case in this branch).

jkoritzinsky commented:
@elinor-fung this is ready for another round of review.

jkoritzinsky committed:
Remove trailing references to StressLogAnalyzer.Tests

elinor-fung commented on src/native/managed/cdacreader/Microsoft.Diagnostics.DataContractReader.Contracts/PrintfStressMessageFormatter.cs:
```diff
@@ -0,0 +1,332 @@
+        string FormatVTable(TargetPointer pointer);
+        string FormatStackTrace(TargetPointer pointer);
+    }
+
```
```suggestion
```

elinor-fung commented on src/native/managed/cdacreader/Microsoft.Diagnostics.DataContractReader.Contracts/PrintfStressMessageFormatter.cs:
```diff
@@ -0,0 +1,332 @@
+
+namespace Microsoft.Diagnostics.DataContractReader;
+
+public sealed class PrintfStressMessageFormatter
```
I'm not sure this really belongs in the Contracts assembly. It looks like a helper for use on top of the contracts, which seems more like part of a contract consumer - so the StressLogAnalyzer?

jkoritzinsky replied:
My plan was to reuse this in other places that need to also implement stress message formatting, like an eventual implementation of SOS's DumpLog. It would need the exact same implementation.

elinor-fung replied:
Okay. I still think it is a bit separate from the Contracts assembly, but I can see this as a reasonable place for now to avoid an entirely separate assembly and package to allow others to consume just for the one class. I think it is worth revisiting if we end up with a bunch more helpers worth reusing on top of the actual contracts.

jkoritzinsky replied:
I'm fine revisiting this later and possibly adding another assembly or something.

jkoritzinsky commented on src/native/managed/cdacreader/Microsoft.Diagnostics.DataContractReader.Contracts/Contracts/StressLog.cs:
```diff
@@ -0,0 +1,335 @@
+    }
+}
+
+#pragma warning disable CS9107 // Parameter 'Target target' is captured into the state of the enclosing type and its value is also passed to the base constructor. The value might be captured by the base class as well.
```
We don't have an explicit property on the base class. I'll see if I can clean this up to not need the pragma and still be in a style I want.

jkoritzinsky committed:
Use composition instead of inheritance

elinor-fung reviewed (approved):
(Sorry, thought I had already submitted this weeks ago.)
